language/transaction-builder/generator/tests/generation.rs

```diff
@@ -298,6 +298,155 @@ fn test_that_java_code_compiles_and_demo_runs() {
     assert!(output.status.success());
 }
 
+#[test]
+#[ignore]
+fn test_that_csharp_code_compiles_and_demo_runs() {
+    let registry = get_diem_registry();
+    let abis = get_stdlib_script_abis();
+    // Special case this because of what the default tempdir is on a mac
+    // It looks as if the path string might be too long for the dotnet runtime
+    // to execute correctly because you get funny errors that don't occur when
+    // the path string is shorter. So make the temp path shorter and all is good.
+    // Avoids this:
+    // "Unhandled exception. System.IO.FileNotFoundException: Could not load file
+    // or assembly \'Diem.Types, Version=1.0.0.0, Culture=neutral,
+    // PublicKeyToken=null\'. The system cannot find the file specified.\n\n
+    // File name: \'Diem.Types, Version=1.0.0.0, Culture=neutral,
+    // PublicKeyToken=null\'\n\n\n"`,
+    if std::env::consts::OS == "macos" {
+        std::env::set_var("TMPDIR", "/private/tmp/");
+    }
+    let dir = tempdir().unwrap();
+
+    let paths = std::fs::read_dir("examples/csharp/custom_diem_code")
+        .unwrap()
+        .map(|e| e.unwrap().path());
+    let config = serdegen::CodeGeneratorConfig::new("Diem.Types".to_string())
+        .with_encodings(vec![serdegen::Encoding::Bcs])
+        .with_custom_code(buildgen::read_custom_code_from_paths(
+            &["Diem", "Types"],
+            paths,
+        ));
+    let bcs_installer = serdegen::csharp::Installer::new(dir.path().to_path_buf());
+    bcs_installer.install_module(&config, &registry).unwrap();
+    bcs_installer.install_serde_runtime().unwrap();
+    bcs_installer.install_bcs_runtime().unwrap();
+
+    let abi_installer = buildgen::csharp::Installer::new(dir.path().to_path_buf());
+    abi_installer
+        .install_transaction_builders("Diem.Stdlib", &abis)
+        .unwrap();
+
+    std::fs::create_dir(dir.path().join("Demo")).unwrap();
+    std::fs::copy(
+        "examples/csharp/StdlibDemo.cs",
+        dir.path().join("Demo/StdlibDemo.cs"),
+    )
+    .unwrap();
+
+    let mut proj =
+        std::fs::File::create(dir.path().join("Diem/Stdlib/Diem.Stdlib.csproj")).unwrap();
+    write!(
+        proj,
+        r#"
+<Project Sdk="Microsoft.NET.Sdk">
+    <ItemGroup>
+        <ProjectReference Include="..\Types\Diem.Types.csproj" />
+    </ItemGroup>
+    <PropertyGroup>
+        <TargetFramework>netstandard1.6</TargetFramework>
+        <LangVersion>7.2</LangVersion>
+    </PropertyGroup>
+</Project>
+"#
+    )
+    .unwrap();
+
+    let status = Command::new("dotnet")
+        .arg("new")
+        .arg("sln")
+        .arg("-n")
+        .arg("Demo")
+        .arg("-o")
+        .arg(dir.path().join("Demo"))
+        .status()
+        .unwrap();
```

Comment:
This is where the codegen error is coming from. My guess is that the FileNoutFound error is coming from the `dotnet` command not being found as opposed to this line `.arg(dir.path().join("Demo"))` since the copy on line 341 looks to have been successful where we copied into the `Demo` directory. 

@bmwill What is the best way to verify?

(oh and it works on my machine 🤣 )

```
---- test_that_csharp_code_compiles_and_demo_runs stdout ----
thread 'test_that_csharp_code_compiles_and_demo_runs' panicked at 'called `Result::unwrap()` on an `Err` value: Os { code: 2, kind: NotFound, message: "No such file or directory" }', language/transaction-builder/generator/tests/generation.rs:373:10
stack backtrace:
   0: rust_begin_unwind
             at /rustc/cb75ad5db02783e8b0222fee363c5f63f7e2cf5b/library/std/src/panicking.rs:493:5
   1: core::panicking::panic_fmt
             at /rustc/cb75ad5db02783e8b0222fee363c5f63f7e2cf5b/library/core/src/panicking.rs:92:14
   2: core::option::expect_none_failed
             at /rustc/cb75ad5db02783e8b0222fee363c5f63f7e2cf5b/library/core/src/option.rs:1268:5
   3: core::result::Result<T,E>::unwrap
             at /rustc/cb75ad5db02783e8b0222fee363c5f63f7e2cf5b/library/core/src/result.rs:973:23
   4: generation::test_that_csharp_code_compiles_and_demo_runs
             at ./tests/generation.rs:365:18
   5: generation::test_that_csharp_code_compiles_and_demo_runs::{{closure}}
             at ./tests/generation.rs:303:1
   6: core::ops::function::FnOnce::call_once
             at /rustc/cb75ad5db02783e8b0222fee363c5f63f7e2cf5b/library/core/src/ops/function.rs:227:5
   7: core::ops::function::FnOnce::call_once
             at /rustc/cb75ad5db02783e8b0222fee363c5f63f7e2cf5b/library/core/src/ops/function.rs:227:5
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.


failures:
    test_that_csharp_code_compiles_and_demo_runs

test result: FAILED. 4 passed; 1 failed; 0 ignored; 0 measured; 1 filtered out; finished in 6.73s
```

Reply:
So looking at how CI is setup it should be getting installed. You could check by running a simple dotnet --help command and seeing if it fails/is able to print out help information.

Reply:
@bmwill Yeah, it was on the `dotnet` command. The `--help` command failed.

```
---- test_that_csharp_code_compiles_and_demo_runs stdout ----
thread 'test_that_csharp_code_compiles_and_demo_runs' panicked at 'called `Result::unwrap()` on an `Err` value: Os { code: 2, kind: NotFound, message: "No such file or directory" }', language/transaction-builder/generator/tests/generation.rs:368:10
stack backtrace:
   0: rust_begin_unwind
             at /rustc/cb75ad5db02783e8b0222fee363c5f63f7e2cf5b/library/std/src/panicking.rs:493:5
   1: core::panicking::panic_fmt
             at /rustc/cb75ad5db02783e8b0222fee363c5f63f7e2cf5b/library/core/src/panicking.rs:92:14
   2: core::option::expect_none_failed
             at /rustc/cb75ad5db02783e8b0222fee363c5f63f7e2cf5b/library/core/src/option.rs:1268:5
   3: core::result::Result<T,E>::unwrap
             at /rustc/cb75ad5db02783e8b0222fee363c5f63f7e2cf5b/library/core/src/result.rs:973:23
   4: generation::test_that_csharp_code_compiles_and_demo_runs
             at ./tests/generation.rs:365:18
   5: generation::test_that_csharp_code_compiles_and_demo_runs::{{closure}}
             at ./tests/generation.rs:303:1
   6: core::ops::function::FnOnce::call_once
             at /rustc/cb75ad5db02783e8b0222fee363c5f63f7e2cf5b/library/core/src/ops/function.rs:227:5
   7: core::ops::function::FnOnce::call_once
             at /rustc/cb75ad5db02783e8b0222fee363c5f63f7e2cf5b/library/core/src/ops/function.rs:227:5
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```